### DISCUSSION TITLE
Supporting functions (aka plugins)

### DISCUSSION THREAD

**Discussion by n3d1117** [n3d1117/chatgpt-telegram-bot#345] [Ideas]
# Help me in adding useful functions to the bot!

## What are functions?
Check out [the blog post](https://openai.com/blog/function-calling-and-other-api-updates) announcing the feature. TLDR:
> Developers can now describe functions to gpt-4-0613 and gpt-3.5-turbo-0613, and have the model intelligently choose to output a JSON object containing arguments to call those functions. This is a new way to more reliably connect GPT's capabilities with external tools and APIs.

## Current implementation
PoC of a few plugins:

<p align="center"><img width="600" alt="image" src="https://github.com/n3d1117/chatgpt-telegram-bot/assets/11541888/657b1804-fbe4-4d98-be0e-22167bdd2f78"></p>

Check out the [`feature/support-functions`](https://github.com/n3d1117/chatgpt-telegram-bot/tree/feature/support-functions) branch containing initial support for functions and some initial functions:
- Daily weather and 7-day forecast for any location (powered by [Open-Meteo](https://open-meteo.com))
- Live cryptocurrencies rate (powered by [CoinCap](https://coincap.io)) - by [@stumpyfr](https://github.com/stumpyfr)
- WolframAlpha queries (powered by [WolframAlpha](https://www.wolframalpha.com)) - requires a WolframAlpha API key in your env variables (`WOLFRAM_APP_ID`)
- Spotify search and personal top tracks/artists - requires `SPOTIFY_CLIENT_ID`, `SPOTIFY_REDIRECT_URI`, `SPOTIFY_CLIENT_SECRET`
- Web search (powered by [DuckDuckGo](https://duckduckgo.com))
- Text translation in any language (powered by [DuckDuckGo](https://duckduckgo.com))
- Image and GIF search (powered by [DuckDuckGo](https://duckduckgo.com))
- World time (powered by [WorldTimeAPI](http://worldtimeapi.org)) - by [@noriellecruz](https://github.com/noriellecruz)
- Dice, simply send a dice in the chat
- YouTube audio extractor (powered by [putube](https://pytube.io/en/latest/))
- DeepL translate - by [@LedyBacer](https://github.com/LedyBacer)

See the full list in the README.

To enable any of these plugins, set them in the `PLUGINS` variable as a comma separated list (e.g: `PLUGINS=weather,crypto,wolfram,spotify,ddg_web_search,ddg_image_search,ddg_translate`).

Set `SHOW_PLUGINS_USED` to `true` to show which plugins were used in generating a response, if any.

Some key points:
- **IMPORTANT**: Adding functions will increase the token usage!
  > Under the hood, functions are injected into the system message in a syntax the model has been trained on. This means functions count against the model's context limit and are billed as input tokens. If running into context limits, we suggest limiting the number of functions or the length of documentation you provide for function parameters. ([source](https://platform.openai.com/docs/guides/gpt/function-calling))
- Functions are currently available only the latest models (`gpt-3.5-turbo-0613`, `gpt-3.5-turbo-16k`, `gpt-3.5-turbo-16k-0613`, `gpt-4-0613`, `gpt-4-32k-0613`). Stable models (`gpt-3.5-turbo`, `gpt-4`, `gpt-4-32k`) will be updated to support functions on June 27, 2023. If you want to try out functions right now, specify one of the new models (e.g. `OPENAI_MODEL=gpt-3.5-turbo-0613`)
- ChatGPT will automatically detect the need for a function call and automatically provide arguments to populate API calls. The JSON response is then fed back to the model and presented to the user in a friendly way. 
- The model might make more than one call to the function while generating a response (for example, if i ask the weather for multiple cities). I've added a new environment variable called `FUNCTIONS_MAX_CONSECUTIVE_CALLS` to control the maximum number of consecutive calls to be made across all functions when generating a single response. It currently defaults to 10.
- All it takes to add a plugin to the chat model is to define a function to fetch the data and its specification. Check out the [`plugins/crypto.py`](https://github.com/n3d1117/chatgpt-telegram-bot/blob/feature/support-functions/bot/plugins/crypto.py) file for a quick example. If you want to add your own function, simply create a .py file in the `plugins` folder subclassing the `Plugin` abstract class, and update the [`functions.py`](https://github.com/n3d1117/chatgpt-telegram-bot/blob/feature/support-functions/bot/functions.py) file accordingly by registering your own plugin. Don't forget to enable it in your .env file, e.g `PLUGINS=your-own-plugin`

## Going forward
The plan is to add many more useful functions to the bot before drafting a new release. Your help is appreciated! 

If you have any questions, cool ideas, utilities, services or APIs that you'd like to see integrated, please comment here or open a PR to the [`feature/support-functions`](https://github.com/n3d1117/chatgpt-telegram-bot/tree/feature/support-functions) branch!

I'm also looking for feedback regarding token usage. Since functions count as input tokens, the more we add the more requests become more expensive. Should functions be loaded on-demand only (i.e. user first selects a function and then uses it)?

**Comment by Jake-double-one** [comment]
I am uncertain if it is technically possible, but maybe you could implement a web search, so the bot can browse the internet.
This might work with the Google Search API, which features a free contingent every month.

**Reply by n3d1117** [reply]
Hey @Jake-double-one, thanks for your feedback. That's definitely one of the main use cases for plugins and I'm planning on adding it! Do you know how much do these APIs cost? I need to look into this, there are also alternatives like https://serpapi.com, Brave Search API and so on...

**Reply by Jake-double-one** [reply]
Sry for my late reply. I was a bit busy.

When starting with an account in the Google cloud (https://console.cloud.google.com/), I received €280 trial balance. I'm using SuperAGI sporadically, and currently I'm still on €0 costs.
I cannot tell you how much it will cost, but I assume, it will be peanuts. 
The biggest issue is the process of receiving the API because it will be very complicated for "non-IT-people".
So if there are other search APIs, I would suggest using them instead of Google. I would rather pay for it if it were easier to set up.

**Comment by stumpyfr** [comment]
I wanted to add few ideas of function and realized the implem was requesting to add `elif` per function so I decided to switch to a plugin architecture, hope it is ok with you!
#347 

I added a small demo with crypto rate lookup:
![Screenshot 2023-06-15 at 15 18 47](https://github.com/n3d1117/chatgpt-telegram-bot/assets/1849743/0d83e292-b699-4677-96d7-6e8a93aa4e86)

**Reply by n3d1117** [reply]
Hi @stumpyfr that's neat! Thanks for your contribution! While this plugin system looks great, what is the advantage of using it instead of an abstract interface?

**Reply by stumpyfr** [reply]
The real value is to be able to register plugins from other packages, like we are doing with pytest.

so we can imagine to do 
```
pip install chatgpt-bot-crypto
``` 
And this package will install a plugin for the bot.

https://pluggy.readthedocs.io/en/stable/

**Reply by n3d1117** [reply]
Alright thanks, I'll keep this in consideration!

**Comment by jnaskali** [comment]
> Should functions be loaded on-demand only (i.e. user first selects a function and then uses it)?

With expected growth in the number of available functions, I think something like this will be a necessity.

Perhaps functions could be dynamically loaded based on natural keywords? Function developers could list activation words, which (if found in user input) trigger the inclusion of the function in the system message. I imagine such word lists would be relatively short (weather, sunny, rain, umbrella, temperature, etc. for the original example) while enabling freeform requests without unnecessary token use.

This could be combined with custom word-list overrides and a function-level white/blacklist system (include function: always / if keyword present / never) that allows for individual customization through environmental variables.

Great thanks for the development work so far :+1:

**Reply by n3d1117** [reply]
Hey @jnaskali thanks for your feedback!

> Perhaps functions could be dynamically loaded based on natural keywords?

This is not feasible imo. Imagine a plugin that integrates Wolfram Alpha, the list of keywords would be endless. Or a follow up questions about the weather in a city:

```
U: What's the weather in Florence? <-- contains 'weather' keyword, all good, using weather func
A: The weather in Florence is [...]
U: What about Rome? <-- does not contain any weather keyword but should still use the weather func
```

The current model is great at detecting the need of function calling in these kind of contexts, and we should not aim to replace it, imo.

**Comment by stanislavlysenko0912** [comment]
For some reason, I thought it would be more interesting if the plugins will be dynamic.

What I'm getting at, for example - I would like to connect some of my service to GPT, while using the original branch of the bot, what would update to new versions, and so on. 

Make some class that will have basic functionality, from this class will be created "function call plugins". 
For each plugin, a separate file with the class, which will be automatically loaded from a certain folder.
And there will be some basic set of functions that people can extend through Git, as well as their own that they want to keep just for themselves.

And then it will be possible to make 2 classes divided by convention:
1 - The class returns the data to GPT, and it already gives some kind of response
2 - The class performs functions and sends the answer itself, without linking to GPT. (conventionally generate a picture and send it to chat).

I believe that this approach will have more features and be more understandable.
Maybe I'm missing, and the good thing is that GPT needs to give some sort of response after the function call, so that the dialog continues to work properly.

How do you look at it?
I'd like to hear your opinion

**Reply by n3d1117** [reply]
Hey @stanislavlysenko0912 thanks for sharing your thoughts!

> 2 - The class performs functions and sends the answer itself, without linking to GPT. (conventionally generate a picture and send it to chat)

I agree this would be great. The bot should be able to directly send text/photos/videos/audios/files after a function call

**Comment by jnaskali** [comment]
> And then it will be possible to make 2 classes divided by convention:
>
> 1 - The class returns the data to GPT, and it already gives some kind of response
>
> 2 - The class performs functions and sends the answer itself, without linking to GPT. (conventionally generate a picture and send it to chat).

If I understood you correctly, the second category would be custom bot commands. You can already add these in using these instructions: https://github.com/n3d1117/chatgpt-telegram-bot/discussions/334

**Reply by stanislavlysenko0912** [reply]
That's a little different from what I was talking about.
Conventionally, instead of writing the command /image_(name of service) I could just write "picture through *service*", GPT would choose which service to use, and the plugin would send the picture itself, of course this is the simplest example 

In my opinion it would expand the functionality of such "plugins", conventionally downloading some files or something else.

**Reply by stanislavlysenko0912** [reply]
A simple example, the first thing that came to mind and that I was able to quickly implement
Perhaps this is extra functionality, it all depends on the imagination of those who will make such plugins. 

This is just my suggestion, but the idea that at least move the plugins to dynamic loading from a folder I would definitely like

<img width="300" alt="image" src="https://github.com/n3d1117/chatgpt-telegram-bot/assets/86171768/0233a0e6-e738-4dfc-85af-d60be4f72a3e">

**Reply by n3d1117** [reply]
That's actually a nice idea! Could you share the snippet you used to create the screenshot of the web page?

**Reply by stanislavlysenko0912** [reply]
This was done quickly with Selenium as a little test, sure you want to add his addiction to the bot?
I think that's not the best idea, since some people will still have to download chromedriver.

Although there probably isn't any other way to do it if you only make do with Python dependencies.

**Comment by n3d1117** [comment]
June 17th update:

Improved various things and added more PoC plugins (web search, wolfram alpha, crypto rates).
- Enabled plugins can now be configured in the env file like this: `PLUGINS=weather,crypto,wolfram,web_search`
- Set `SHOW_PLUGINS_USED` to true to show which plugins were used in generating a response
- to use WolframAlpha plugin, you need to set the `WOLFRAM_APP_ID`

**Reply by bentacos** [reply]
for the web_search, does chatgpt automatically use it if it doesn't know the answer to your question?

for example, if you asked "how much did johnwick 4 earned in box office?" will the plugin for web_search be utilized by chatgpt even if you don't asked it to search the internet?

**Reply by n3d1117** [reply]
Yes! Works pretty well (assuming the top results are correct)
<img width="600" alt="image" src="https://github.com/n3d1117/chatgpt-telegram-bot/assets/11541888/57ee8b02-b874-4c4a-9a8a-cd0bed51c064">

**Reply by bentacos** [reply]
im testing it rn, when you don't use a plugin in **PLUGINS**, it will still look for the required plugin fields which causes error

**Reply by n3d1117** [reply]
@bentacos sorry I don't get it, could you elaborate?

**Reply by bentacos** [reply]
for example, i used this:

PLUGINS=weather,crypto,wolfram,web_search

when i run the script, it will show errors looking for the spotify credentials

**Reply by n3d1117** [reply]
@bentacos Fixed! Thanks for reporting

**Comment by n3d1117** [comment]
June 19th update:

Various improvments and added a PoC Spotify plugin. Demo:

<img width="600" alt="image" src="https://github.com/n3d1117/chatgpt-telegram-bot/assets/11541888/2d9bf430-a0b3-49cd-909c-9b50b168988c">

Uses [spotipy](https://spotipy.readthedocs.io/en/2.22.1/). Requires `SPOTIFY_CLIENT_ID`, `SPOTIFY_CLIENT_SECRET` and `SPOTIFY_REDIRECT_URI`. Can fetch user's current playing song, user's top artists and tracks in last month, 3 months or years, along with search ability (by query and by id). Works for songs, artists and albums.

**Reply by bentacos** [reply]
so, if a user asks "what is the song currently playing?" will they get a result from the admin's Spotify account?

**Reply by n3d1117** [reply]
as of right now, yeah

**Reply by JayGwod** [reply]
When I use spotify, it prompts "Paste that url you were directed to to complete the authorization", but it doesn't give me a chance to paste that url and then throw an error directly. How to fix this problem?

**Comment by n3d1117** [comment]
June 22nd update:

- Added image search and translate plugins (all powered by DuckDuckGo)
- Tested a PoC python interpreter plugin, works great but it's way too dangerous imo (could possibly read/modify the filesystem...)

**Reply by noriellecruz** [reply]
is the interpreter included in the branch? i would like to test it.

**Reply by n3d1117** [reply]
@noriellecruz it's not, but here's the plugin I originally wrote:

```python
class PythonPlugin(Plugin):
    """
    A plugin to act as a python interpreter
    """
    def get_source_name(self) -> str:
        return "Python"

    def get_spec(self) -> [Dict]:
        return [{
            "name": "python",
            "description": "Run python code. Refuse to run any dangerous code that touches the filesystem.",
            "parameters": {
                "type": "object",
                "properties": {
                    "code": {"type": "string", "description": "Code to run. Should end with a print statement"}
                },
                "required": ["code"],
            },
        }]

    async def execute(self, function_name, **kwargs) -> Dict:
        code = kwargs['code']
        try:
            sys.stdout = StringIO()
            exec(code, globals())
            output = sys.stdout.getvalue().strip()

            if not output:
                lines = code.strip().splitlines()
                last_line = lines[-1]
                last_expression = eval(last_line, globals(), locals())
                output = str(last_expression)

            return {"output": output}

        except Exception as e:
            return {"error": str(e)}

        finally:
            sys.stdout = sys.__stdout__
```

**Comment by ChargerMc** [comment]
Excelente work @n3d1117! May I ask if we provide a link to a PDF on the prompt, its possible that the bot read it a response with information about the PDF?

**Reply by n3d1117** [reply]
Yes, I was planning on building this! The hardest part is finding some free time 😃 
In case someone wants to help, there's a great notebook that covers this topic: https://github.com/openai/openai-cookbook/blob/28fef26cc9df2ccf947c046031d7f0430d1ad8f3/examples/How_to_call_functions_for_knowledge_retrieval.ipynb

**Reply by ChargerMc** [reply]
I would like to help you, but i dont know python right now :c , im going to study this months, maybe i can help later. Thanks for the great job

**Comment by noriellecruz** [comment]
i added a plugin to get time based on location or timezone #360 

![image](https://github.com/n3d1117/chatgpt-telegram-bot/assets/46517631/737e4ceb-0839-4f43-8b4a-d6969f02f4c8)

requires `WORLDTIME_DEFAULT_TIMEZONE` in .env file so the bot will use a default timezone if the user did not specify one, get tz identifiers from here: [https://en.wikipedia.org/wiki/List_of_tz_database_time_zones](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones)

**Comment by n3d1117** [comment]
July 2nd update:
- Added World Time plugin by @noriellecruz 
- `web_search` is now `ddg_web_search`, `image_search` is now `ddg_image_search`, `translate` is now `ddg_translate`
- Some minor fixes and laid the ground work for direct responses

**Comment by n3d1117** [comment]
July 3rd update:
- Added `DUCKDUCKGO_SAFESEARCH` optional env var (`on`, `off`, `moderate`)
- DuckDuckGo searches (images and web) now also take the region in consideration
- Direct responses are now available! You can now send a response from a plugin directly to the chat without first re-feeding it back to the model, as requested by @stanislavlysenko0912 
    - `ddg_image_search` plugin now returns gif/images directly in the chat!
- Added DeepL translate plugin by @LedyBacer
- Added YouTube audio extractor plugin - uses [pytube](https://pytube.io/en/latest/), see [this](https://github.com/pytube/pytube/issues/1707#issuecomment-1616464219) if you have issues like I did
- Added dice plugin to send a dice to the chat

Known limitations:
- Sending direct responses is not possible in inline mode (as far as I know)
- DIrect responses currently don't count towards the used tokens

---

Example of direct response:
```python
# In your plugin
async def execute(self, function_name, **kwargs) -> Dict:
    # do your stuff here
    return {
        'direct_result': {
            'kind': 'photo', # currently supported: photo, gif, file, dice
            'format': 'url',  # currently supported: url, path
            'value': 'http://...' # the actual value for the content. can be a string, a url or a path to a local file
        }
    }
```

Currently supported direct responses:
- Photo from URL: `{'kind': 'photo', 'format': 'url', 'value': 'http://example.com/file.jpg'}`
- Photo from file: `{'kind': 'photo', 'format': 'path', 'value': '/path/to/file.jpg'}`
- Gif from URL: `{'kind': 'gif', 'format': 'url', 'value': 'http://example.com/file.gif'}`
- Gif from file: `{'kind': 'gif', 'format': 'path', 'value': '/path/to/file.gif'}`
- Any remote file: `{'kind': 'file', 'format': 'url', 'value': 'http://example.com/file.pdf'}` - file will be sent as a document
- Any local file: `{'kind': 'file', 'format': 'path', 'value': '/path/to/file'}` - file will be sent as a document and automatically deleted afterwards
- A dice: `{'kind': 'dice', 'format': 'dice', 'value': '🎲'}`

Check out the `handle_direct_result` function in `utils.py` to see how these are handled.
Feel free to add more and give feedback!

---

Demo:
![image](https://github.com/n3d1117/chatgpt-telegram-bot/assets/11541888/487abdb3-b4e0-4480-9b27-ee13204d195d)

**Reply by Celegans12** [reply]
Audio extractor doesnt work sometimes because the titles of some videos use symbols that are invalid as a filename. 
Fixed it at line 36: output = re.sub(r'[^\w\-_\. ]', '_', video.title) + '.mp3'
You also need to import re at the start.
It was also set to save as .mp4 so I changed that to .mp3

**Reply by n3d1117** [reply]
Fixed! Thanks @Celegans12

**Comment by n3d1117** [comment]
July 5th update:
- Added text-to-speech plugin using [gtts](https://pypi.org/project/gTTS/)
- Fix audio extractor plugin not working for some videos
- Removed plugin dependencies from the `requirements.txt` file
    - They can now be installed separately. Added a `Dependency` column to the README plugins table

**Comment by Annagrenadin** [comment]
Hello! I just admire you! Thank you for your work! Please tell me how I can change the following things in the bot: 1. i cannot change the first prompt. I enter my own version in the env file, but for some reason it doesn't work. and 2 question: how can I remove the initial message that lists the commands? 
![image](https://github.com/n3d1117/chatgpt-telegram-bot/assets/139069701/fa12acba-58d8-4ef6-8cb2-130e28420b87)

I need so that a person starts the bot, and the answer to the given prompt immediately comes out?

**Reply by noriellecruz** [reply]
hire someone to do that for you

**Reply by n3d1117** [reply]
Hi @Annagrenadin apologies for the late reply.
1. You mean the assistant prompt? You can change it in .env file like this: `ASSISTANT_PROMPT="You now talk like Yoda"` and it should work after a restart
2. Untested, but you could replace the help function with this:

```python
async def help(self, update: Update, _: ContextTypes.DEFAULT_TYPE) -> None:
  with update.message._unfrozen() as message:
      message.text = "Hello" # or anything, really
  await self.prompt(update=update, context=context)
```

**Reply by Annagrenadin** [reply]
Thank you very much!❤️❤️❤️

**Comment by Annagrenadin** [comment]
![image](https://github.com/n3d1117/chatgpt-telegram-bot/assets/139069701/537057d2-29c1-402d-8fc1-6397429f1743)
I will also be very grateful to you if you tell me how to embed this function in the bot

**Reply by n3d1117** [reply]
Sure, just run `pip install gtts~=2.3.2`, then add `PLUGINS=gtts_text_to_speech` to your `.env` file and finally ask the bot something along the lines of `Send me an audio saying I'm cool`

**Comment by SLV54** [comment]
Great job @n3d1117. I like it. Is it possible to add and train a chatbot on your own thematic knowledge base (for example: text files, websites, FAQs). Maybe there is a script for such a specific bot?

**Reply by n3d1117** [reply]
It's definitely possible! Check out https://www.chatpdf.com and https://vault.pash.city

**Comment by n3d1117** [comment]
July 28th update:

- Added whois plugin by @jnaskali

**Comment by tsymbaluyk** [comment]
Are you planning to add the ability to retrain the bot based on data from the Internet or downloaded files?

**Reply by n3d1117** [reply]
It's possible, but not planned for the upcoming release

**Comment by AlyxAbyss** [comment]
> ![image](https://user-images.githubusercontent.com/139069701/252160138-537057d2-29c1-402d-8fc1-6397429f1743.png)

Seems to work well. Only the default voice in German sounds very artificial. Is there a possibility to use a different gtts voice?

**Reply by n3d1117** [reply]
It does not, as far as I know. Here's the [gtts repo](https://github.com/pndurette/gTTS)
